crates/pipeline-manager/src/cluster_health.rs

```diff
@@ -0,0 +1,164 @@
+use chrono::{DateTime, Utc};
+use log::{info, warn};
+use reqwest::Client;
+use serde::Serialize;
+use std::{error::Error, sync::Arc, time::Duration};
+use tokio::sync::RwLock;
+use utoipa::ToSchema;
+
+/// Interval between checking health again.
+const POLL_INTERVAL: Duration = Duration::from_secs(10);
+
+/// Default HTTP request timeout to use
+const DEFAULT_REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Clone, Serialize, PartialEq, ToSchema, Default)]
+pub struct HealthStatus {
+    pub runner: ServiceStatus,
+    pub compiler: ServiceStatus,
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, ToSchema)]
+pub struct ServiceStatus {
+    pub healthy: bool,
+    pub message: String,
+    pub unchanged_since: DateTime<Utc>,
+}
+
+impl Default for ServiceStatus {
+    fn default() -> Self {
+        Self {
+            healthy: false,
+            message: String::new(),
+            unchanged_since: Utc::now(),
+        }
+    }
+}
+
+async fn poll_service(url: &str, client: &Client, timeout: Duration) -> (bool, String) {
```

Comment:
Could you make the error message more informational/actionable and formatted? As in:
```
The latest health check timed out. This indicates that ... . Please check the logs of the runner for additional information about the cause.
```

Along those lines, maybe add another field called request_error or so for a more technical error. The message can also be optional, as in the healthy case there is no supplemental information.

Reply:
I want to avoid every client coming up with their own nicely formatted message, the back-end should give extra context that then can be literally displayed.

Reply:
fda, Web Console, Python, etc. when they are used to monitor cluster health and need to report an error about what is going on.